### PR TITLE
Refactor: Unify the interfaces of Pulay terms of force and stress

### PR DESCRIPTION
**Pull request by maki49**
### Why need this change
To implement the analytical gradient of LR-TDDFT, we need to calculate:
image

where $(T+D^{Z})$ is a density matrix different from the ground-state (GS) one, and $\Lambda$ is a complicated energy density matrix different from the GS one.

However in the current code, the GS energy density matrix is hard-coded in `cal_fedm`, blocking me from implementing the feature. 

Additionally, I notice that if the energy density matrix is treated equally as the density matrix as the input argument, the Pulay terms of force can be abstracted to $\text{Tr}[D\frac{dH}{dr_x}]$, and similarly $\frac{1}{\Omega}\text{Tr}[D\frac{dH}{dr_x}r_y]$ for stress, which can avoid a lot of duplicate code (`cal_fedm`, `cal_ftvnl_dphi` and `cal_fvnl_dphi`).

### What's changed
Unify `cal_fedm`, `cal_ftvnl_dphi` and `cal_fvnl_dphi` (both gamma-only and multi-k) into a new function `cal_pulay_fs` , where 3 overloads are provided:
- calculated by 2-center integration, provided dH(x, y, z) and dH*r(11, 12, 13, 21, 22, 23)
- calculated by 2-center integration, provided dH(x, y, z) and dr(x,y,z)
- calculated by grid integration

All of them above treat the (energy) density matrix as an input parameter, and the Hamiltonian term (any operator) is passed in the way according to the algorithm.

### CODE REVIEW

File: source/Makefile.Objects [hunk 1]
```diff
@@ -56,6 +56,7 @@ VPATH=./src_global:\
 ./module_hamilt_lcao/module_dftu:\
 ./module_hamilt_lcao/module_deltaspin:\
 ./module_hamilt_lcao/hamilt_lcaodft/operator_lcao:\
+./module_hamilt_lcao/hamilt_lcaodft/pulay_force_stress:\
```
Comment (mohanchen) [inline source/Makefile.Objects]: don't randomly add new directory in the code, if everyone work in this way, the files will be organlized in a messy way

**Comment by dyzheng** [comment]
The refactor of force or stress code should avoid the memory cost of $$\frac{dH}{d\tau_i}$$ for force or $$\frac{dH}{d\tau_\alpha}\tau_\beta$$ for stress, otherwise it will be refactored again in the future.

**Comment by maki49** [comment]
> The refactor of force or stress code should avoid the memory cost of d H d τ i for force or d H d τ α τ β for stress, otherwise it will be refactored again in the future.

You can delete those arrays in `fsr` later. It does not conflict with my demands.

**Comment by mohanchen** [comment]
> You can delete those arrays in `fsr` later

In PRs, try to fix the issues proposed by others, don‘t let others do anything to meet your demands. Nobody has the obligation to do something to meet your personal demands. You need to consider more for others.

**Comment by maki49** [comment]
> In PRs, try to fix the issues proposed by others, don‘t let others do anything to meet your demands. Nobody has the obligation to do something to meet your personal demands. You need to consider more for others.

I didn't let others do anything to meet my demands. I'm meeting my demands **by myself**. 
Deleting `fsr` (already existing, not my addition) is another plan of @dyzheng. We've talked about it.
The two things do not conflict.

File: source/module_hamilt_lcao/hamilt_lcaodft/force_stress_arrays.h [hunk 2]
```diff
@@ -60,6 +60,10 @@ class ForceStressArrays
     double* DSloc_23;
     double* DSloc_33;
 
+    //缺少DSlocR_11, DSlocR_12, DSlocR_13, DSlocR_22, DSlocR_23, DSlocR_33
+    //DH_r更省内存
+    // 数组接口需要重构
```
Comment (mohanchen) [inline source/module_hamilt_lcao/hamilt_lcaodft/force_stress_arrays.h]: use English comments?

**Review by mohanchen — APPROVED** [review]

**Pull request merged**